packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx

```diff
@@ -153,32 +153,46 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
   const toggleButtonRef = React.useRef<HTMLSpanElement>(null);
   const mounted = React.useRef(false);
 
+  // Record to store all extra information (e.g. original indexes), where the key is the id of the option
+  const optionsExtraInfoRef = React.useRef<
+    Record<string, { __originalIndex: number }>
+  >({});
+
+  // Function to initialize extra info
+  const initializeExtraInfo = React.useCallback(() => {
+    optionsProp.current.forEach((option, index) => {
+      optionsExtraInfoRef.current[getOptionId(option, id)] = {
+        __originalIndex: index,
+      };
+    });
+  }, [id]);
+
   // Latest value of the onChange prop
   const onChangeProp = React.useRef(onChange);
   React.useEffect(() => {
     onChangeProp.current = onChange;
   }, [onChange]);
 
-  // Record to store all extra information (e.g. original indexes), where the key is the id of the option
-  const optionsExtraInfoRef = React.useRef<
-    Record<string, { __originalIndex: number }>
-  >({});
+  // Latest value of the options prop
+  const optionsProp = React.useRef(options);
 
   // Clear the extra info when the options change so that it can be reinitialized below
   React.useEffect(() => {
     optionsExtraInfoRef.current = {};
-  }, [options]);
+  }, [optionsProp]);
```

Comment:
What is the difference between `options` and `optionsProp`?

Reply:
`options` is array of options user provides.
`optionsProp` is mutable ref object we create.
Having `optionsProp` helps us to control when we change options and selected value in combobox, helps us prevent infinite bouncing loops between useEffects.
You can read more about React hooks [here](https://reactjs.org/docs/hooks-reference.html#useref).